SHiRA Adapters

kkb-code opened the pull request:
We would like to add code for Sparse High Rank Adapters (SHiRA) which was published at NeurIPS 2024 ([PAPER LINK](https://neurips.cc/virtual/2024/poster/96503)). 

This is an alternate type of adapter and has been found to have significant advantages over the low rank adapters. Specifically, SHiRA achieves better accuracy than LoRA for a variety of vision and language tasks. It also offers simpler and higher quality multi-adapter fusion by significantly reducing concept loss, a common problem faced by low rank adapters. Please see the paper for more details.

kkb-code commented:
Issue: #2585

BenjaminBossan reviewed (changes requested):
Thanks for this PR to integrate SHiRA into PEFT. It's good to see a paper that not only deals with academic improvements but also considers practical applications.

I went through the code and I think this generally already looks quite fine. Still, I found a couple of issues, please check.

Regarding testing: I saw that you added `test_shira.py`. This is fine for SHiRA-specific tests, but for the general PEFT integration, we need to update the existing PEFT tests to include SHiRA. As a first step, let's add one or a few examples using SHiRA to the custom model tests here:

https://github.com/huggingface/peft/blob/759bb70acea8486b3a493792a5b6ba80b5217fb4/tests/test_custom_models.py#L68

After that, please run `pytest tests/test_custom_models.py -k "shira"` to see if these tests pass.

Ideally, before we merge, we should also:

1. add docs
2. add an example
3. add a [benchmark](https://github.com/huggingface/peft/tree/main/method_comparison/MetaMathQA)

Those steps are not strictly necessary, as they can be added later, but often it's a good idea to include them, as they can sometimes reveal new areas for improvement. But it's fine to focus on the proper implementation for now.

Finally, please always `run style` before pushing your changes to ensure that the linter is happy.

kkb-code commented on src/peft/tuners/shira/__init__.py:
```diff
@@ -0,0 +1,27 @@
+register_peft_method(name="shira", config_cls=ShiraConfig, model_cls=ShiraModel, prefix="shira_", is_mixed_compatible=True)
+
+
+def __getattr__(name):
```
This is done.

kkb-code commented on src/peft/tuners/shira/layer.py:
```diff
@@ -0,0 +1,194 @@
+            return
+        self.scaling[adapter] = scale
+
+class Linear(nn.Linear, ShiraLayer):
```
Currently SHiRA is only supported for linear layers only. This is why we used nn.Linear and not nn.Module since the code depends on input and output features to compute num_shira_weights. We followed VeRA code that is also supported for linear layers. Please see: [Linear class in VeRA code](https://github.com/huggingface/peft/blob/main/src/peft/tuners/vera/layer.py#L141)

BenjaminBossan replied:
Hmm, not quite sure if I follow, you can use `self.in_features` and `self.out_features`. I'd prefer to have consistent types here with other PEFT methods, which use `nn.Module`. There could be `isinstance(layer, nn.Linear)` calls somewhere that could be tripped up by the changed inheritance structure.

kkb-code commented on src/peft/tuners/shira/layer.py:
```diff
@@ -0,0 +1,194 @@
+        elif self.merged:
+            result = self.base_layer(x, *args, **kwargs)
+        else:
+            new_weight = copy.deepcopy(self.base_layer.weight.data)
```
One of the problems we faced was that self.base_layer.weight had requires_grad set to False (since it is a base weight). If we scatter something on top of a base_layer.weight and try to use it inside F.linear(...), the gradients were not flowing to SHiRA weights. We are open to suggestions to improve this.

BenjaminBossan replied:
Thinking more about this, I would really prefer if we would call `result = self.base_layer(x, *args, **kwargs)` and then add the shira activations on top of the base result, instead of modifying the weight and calling `F.linear`. This is because we should not assume that the base layer simple calls `F.linear`, there could be other things going on in the `forward` call that we're missing like this (could be an `nn.Linear` subtype, could have forward hooks, ...).

So I came up with this revised `forward` method, could you please check if that is correct?

```python
    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
        if self.disable_adapters:
            if self.merged:
                self.unmerge()
            result = self.base_layer(x, *args, **kwargs)
        elif self.merged:
            result = self.base_layer(x, *args, **kwargs)
        else:
            result = self.base_layer(x, *args, **kwargs)
            torch_result_dtype = result.dtype
            delta_weight = None
            for active_adapter in self.active_adapters:
                if active_adapter not in self.shira_weight.keys():
                    continue

                if delta_weight is None:
                    delta_weight = self.get_delta_weight(active_adapter)
                else:
                    delta_weight = delta_weight + self.get_delta_weight(active_adapter)

            if delta_weight is not None:
                x = self._cast_input_dtype(x, delta_weight.dtype)
                result = result + F.linear(x, delta_weight).to(torch_result_dtype)

        return result
```

kkb-code commented on src/peft/tuners/shira/mask_functions.py:
```diff
@@ -0,0 +1,41 @@
+"""
+
+def random_mask(base_layer: nn.Module, r: int, random_seed: int = 42, **kwargs) -> torch.tensor:
+    _ = kwargs
```
This is done.

kkb-code commented on src/peft/tuners/shira/mask_functions.py:
```diff
@@ -0,0 +1,41 @@
+    shape = base_layer.weight.shape
+    num_shira_weights = r * (shape[0] + shape[1])
+    random_generator = torch.Generator()
+    random_generator.manual_seed(random_seed)
```
We have fixed this now by making the default seed as None. Users can supply a specific seed if they want to reproduce their results for a specific experiment.

kkb-code commented on src/peft/tuners/shira/layer.py:
```diff
@@ -0,0 +1,194 @@
+        self.shira_embedding_weights = nn.ParameterDict({})
+        self.weight_shape = base_layer.weight.shape #Assumes SHiRA is on some layer with "weight" parameter
+        self.device = base_layer.weight.device
+        self.num_shira_weight = {}
```
We have fixed this now.

kkb-code commented on tests/test_shira.py:
```diff
@@ -0,0 +1,328 @@
+# Copyright 2024-present the HuggingFace Inc. team.
```
We have fixed this now.

kkb-code commented on tests/test_shira.py:
```diff
@@ -0,0 +1,328 @@
+        assert output.dtype == dtype
+
+
+if __name__ == '__main__':
```
We have fixed this now.

BenjaminBossan replied:
I mean please remove the whole `if __name__ == '__main__':` part, it is not necessary as we use pytest.

kkb-code commented on src/peft/tuners/shira/config.py:
```diff
@@ -0,0 +1,97 @@
+            with LoRA adapter types in terms of parameter count for direct comparison.  Thus, for a given target module, the number
+            of SHiRA parameters will be set to r(m+d) where the original tensor dimensions are m x d.  This means the number of
+            additional parameters is the same as a LoRA adapter.
+        mask_fn_tag (`str`, defaults to `random_mask`):
+            Name of the mask function. Since the mask_fn is not serializable, this field can be used to keep a record of mask_fn. This string is not used in the implementation.
+        mask_fn (`function`, defaults to `None`):
+            An optional user-defined mask_fn to compute the mask value from base_weights. For pretrained weight with shape m, n, return only one mask (shape: m, n) 
+            which must be binary 0 or 1 with num_shira_parameters = r(m+n) for linear layers. Device and dtype of mask must be same as base layer's weight's device and dtype. 
+            If mask_fn is None, then it will generate a random sparse mask of size r(m + n). This function can be found inside mask_functions.py."
```
We believe that the future users may want to experiment with their own custom mask functions without requiring PEFT library changes. This is why we used a function reference in this field to allow for maximum flexibility (and we suggest that keeping this functionality would be beneficial to the community).

kkb-code commented on src/peft/tuners/shira/layer.py:
```diff
@@ -0,0 +1,194 @@
+
+class ShiraLayer(BaseTunerLayer):
+    # List all names of layers that may contain trainable adapter weights
+    adapter_layer_names = ("shira_weight", "shira_embedding_weights")
```
We wanted to keep this functionality because it seems to be commonly used in other adapter types. However, we did not have time to explore the full implementation for this support. Would you suggest that we remove this? Or let the future contributors build further on this?

BenjaminBossan commented on tests/test_shira.py:
```diff
@@ -0,0 +1,328 @@
+
+        return peft_model
+
+    def test_mlp_single_adapter_merge_unmerge(self, mlp):
```
This functionality should be covered by `test_custom_models.py` and can thus be removed.

kkb-code replied:
Removed.

BenjaminBossan commented on tests/test_shira.py:
```diff
@@ -0,0 +1,328 @@
+
+        return peft_model
+
+    def test_mlp_multiple_adapters_same_attachment_points(self, mlp):
```
This functionality should be covered by `test_custom_models.py` and can thus be removed.

kkb-code replied:
Removed.

BenjaminBossan commented on tests/test_shira.py:
```diff
@@ -0,0 +1,328 @@
+
+        return peft_model
+
+    def test_mlp_multiple_adapters_different_attachment_points(self, mlp):
```
This will be covered by `test_custom_models.py` if you add entries here: https://github.com/huggingface/peft/blob/6bcefb02c66ff6821fbbe58bc4d8fc864e2db335/tests/test_custom_models.py#L549.

kkb-code replied:
Fixed. We added a config argument init_randn_shira_weight in ShiraConfig so that the SHiRA weights are not initialized to zeros. This was done to support and pass many pytests.

BenjaminBossan commented on tests/test_shira.py:
```diff
@@ -0,0 +1,328 @@
+
+        return peft_model
+
+    def test_mlp_multiple_adapters_different_attachment_points_merge_and_unload(self, mlp):
```
This will be covered by `test_custom_models.py` if you add entries here: https://github.com/huggingface/peft/blob/6bcefb02c66ff6821fbbe58bc4d8fc864e2db335/tests/test_custom_models.py#L549.

BenjaminBossan reviewed (changes requested):
Thanks a lot for the updates, this is shaping up quite well.

I still have a couple of comments, but those should not be too hard to address.

Moreover, let's work on the following to finalize the PR:

1. Please check your ruff version. It doesn't look like `make style` worked as expected. The CI uses ruff-0.9.10.
2. Let's add an entry to the docs, check how the other PEFT methods do it.
3. Let's add an example to the `examples/` folder, what exactly you want is up to you. You can take an existing one and modify it for SHiRA.
4. Let's extend the test coverage by adding entries to `test_decoder_models.py`, `test_encoder_decoder_models.py`, `test_feature_extraction_models.py` and `test_seq_classifier.py`. This should be straightforward, as you only need to add a config to the list of test cases.

kkb-code commented:
> Ah, forgot to reply to this:
> 
> > We believe that the future users may want to experiment with their own custom mask functions without requiring PEFT library changes. This is why we used a function reference in this field to allow for maximum flexibility (and we suggest that keeping this functionality would be beneficial to the community).
> 
> IMO, it makes little sense to have a config parameter (`mask_fn`) that isn't persisted. It means that the user needs to run custom code after loading the model to apply the correct mask function. If they do that anyway, we might as well remove the argument. 

Please note that our SHiRA save_and_load.py part saves the indices used during training and loads them correctly during inference. This is so that the users have 100% reproducibility when it comes to using the exact same SHiRA indices in the mask that they trained the adapter on. Regardless, we have now adopted your method of calling the mask function once the SHiRA config is declared.

> I think something like this would work:
> 
> ```python
> @dataclass
> class ShiraConfig(PeftConfig):
>     ...
>     mask_type: Litera["random", <more-options>] = field(default='random', metadata={"help": "..."})
> 
>     def __post_init__(self):
>         if self.mask_type == "random":
>             self.mask_fn = random_mask
>         elif ...
>         else:
>             warnings.warn(f"Argument {mask_type=} not recognized, please supply your own masking function by calling `config.mask_fn = my_mask_fn`.")
>             self.mask_fn = None
> ```
> 
> Then, if users want to provide a custom mask function, they can still do:
> 
> ```python
> shira_config = ShiraConfig(...)
> shira_config.mask_function = my_custom_mask_function
> ```

We agree that this is a much nicer way to support user defined mask functions. Hence, we have added this to our SHiRA implementation.

kkb-code commented:
> Thanks a lot for the updates, this is shaping up quite well.
> 
> I still have a couple of comments, but those should not be too hard to address.
> 
> Moreover, let's work on the following to finalize the PR:
> 
> 1. Please check your ruff version. It doesn't look like `make style` worked as expected. The CI uses ruff-0.9.10.
> 2. Let's add an entry to the docs, check how the other PEFT methods do it.
> 3. Let's add an example to the `examples/` folder, what exactly you want is up to you. You can take an existing one and modify it for SHiRA.
> 4. Let's extend the test coverage by adding entries to `test_decoder_models.py`, `test_encoder_decoder_models.py`, `test_feature_extraction_models.py` and `test_seq_classifier.py`. This should be straightforward, as you only need to add a config to the list of test cases.

We have added an entry in the docs, please let us know if this looks good. We have fixed all ruff related issues on our end by using the correct version. A SHiRA example has also been added. All the tests are now covered. 

There were a few items that we could not respond in-line above but all of those have been fixed also. Please let us know if you need anything else from our side in order to merge it. We did a merge with the main branch also in one of our recent commits to make things easier :)

BenjaminBossan commented on src/peft/tuners/shira/config.py:
```diff
@@ -0,0 +1,129 @@
+        default=False,
+        metadata={"help": "Set this to True if the layer to replace stores weight like (fan_in, fan_out)"},
+    )
+    init_randn_shira_weight: bool = field(
```
Let's rename this to `init_weights` for consistency with other PEFT methods.

kkb-code replied:
> Let's rename this to `init_weights` for consistency with other PEFT methods.

On "init_weights".  We can rename, but we were afraid the new users would expect that this was an intended step prior to training.  Since SHiRA usually initializes to zeros prior to training, this parameter was really only added to support many of the pytests that assume that loading and unloading, and activating and deactivating with the initialized adapter should cause changes in the weights and outputs.  If you still think it is better to name it the same as the others despite this difference in intent, we can do it. Please let us know. Thanks.

BenjaminBossan replied:
Yes, I think renaming makes sense. It is consistent with how other PEFT methods use `init_weights`, so the PEFT users should have the right expectations.

BenjaminBossan commented on tests/test_custom_models.py:
```diff
@@ -1195,6 +1231,9 @@ def test_safe_merge(self, test_name, model_id, config_cls, config_kwargs):
             # VBLoRA do not take init_weights
             pass
+        elif issubclass(config_cls, ShiraConfig):
+            # SHiRA does not take init_weights
```
If `init_randn_shira_weight` is renamed, this exception can be removed.

kkb-code replied:
This is done.

BenjaminBossan commented on src/peft/tuners/shira/config.py:
```diff
@@ -0,0 +1,129 @@
+            )
+        },
+    )
+    mask_type: Literal["random"] = field(
```
WDYT about supporting the other masking functions from the paper?

BenjaminBossan commented on src/peft/tuners/shira/config.py:
```diff
@@ -0,0 +1,129 @@
+
+    Args:
+        r (`int`, *optional*, defaults to `32`):
+            As SHiRA is high rank, it does not really make sense to choose a "rank".  However, we would like a way to be consistent
```
Some of these docstrings could use some work. Put yourself into the shoes of a user who may not exactly know how SHiRA works, something like

> As SHiRA is high rank, it does not really make sense to choose a "rank"

is not so useful. Instead, this is more important info:

> for a given target module, the number of SHiRA parameters will be set to r(m+d) where the original tensor dimensions are m x d.

For each parameter, ensure that the information about what it does and how increasing/decreasing it affects the model are put first. The rest can be added at the end or even removed.

kkb-code replied:
This is done.

kkb-code commented:
> > This is strange. We were using torch==2.5.1 and transformers==4.51.3 and all tests were passing. We have an A100 GPU.
> 
> Could you please try again with recent versions of torch and transformers?

Following the peft gpu Dockerfile, we upgraded torch to 2.7.1 and transformers to 4.54.0.dev0 (transformers was built from source following the peft gpu Dockerfile [here](https://github.com/huggingface/peft/blob/main/docker/peft-gpu/Dockerfile#L56); most other packages from that dockerfile were also installed).

With this new setup, we are still getting some errors for the gemma model (3 tests fail for SHiRA) similar to what you were seeing. However, for this latest peft-gpu setup, these errors are not specific to SHiRA. VeRA and LoRA tests also fail for the Gemma model with similar errors. Please see below for more details:

```
# SHiRA test errors with torch 2.7.1 and transformers 4.54.0.dev0.
=============================================================================================================================================== short test summary info ================================================================================================================================================
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate[VeraConfig-config_kwargs15-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpzm9mve_1/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpzm9mve_1/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate_pos_args[VeraConfig-config_kwargs15-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpcveqxqm3/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpcveqxqm3/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_disable_adapter[VeraConfig-config_kwargs15-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpivfoy3wp/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpivfoy3wp/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
================================================================================================================ 3 failed, 513 passed, 199 skipped, 13259 deselected, 402 warnings in 341.43s (0:05:41) ================================================================================================================
```

```
#  VeRA test errors with torch 2.7.1 and transformers 4.54.0.dev0.
=============================================================================================================================================== short test summary info ================================================================================================================================================
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate[VeraConfig-config_kwargs15-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpzm9mve_1/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpzm9mve_1/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate_pos_args[VeraConfig-config_kwargs15-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpcveqxqm3/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpcveqxqm3/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_disable_adapter[VeraConfig-config_kwargs15-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpivfoy3wp/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpivfoy3wp/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
================================================================================================================ 3 failed, 513 passed, 199 skipped, 13259 deselected, 402 warnings in 341.43s (0:05:41) ================================================================================================================
```

```
# LoRA test errors with torch 2.7.1 and transformers 4.54.0.dev0.
=============================================================================================================================================== short test summary info ================================================================================================================================================
FAILED tests/test_decoder_models.py::TestDecoderModels::test_mixed_adapter_batches[LoraConfig-config_kwargs7-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpj7ewyq28/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpj7ewyq28/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_mixed_adapter_batches[LoraConfig-config_kwargs8-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpriouz6a1/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpriouz6a1/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate[AdaLoraConfig-config_kwargs0-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmp4vllqb0i/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmp4vllqb0i/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate[LoraConfig-config_kwargs7-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpavw8qucd/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpavw8qucd/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate[LoraConfig-config_kwargs8-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpj07pamco/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpj07pamco/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate[VBLoRAConfig-config_kwargs14-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._dynamo.exc.Unsupported: Data-dependent branching
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate_pos_args[AdaLoraConfig-config_kwargs0-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpckzzm8z1/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpckzzm8z1/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate_pos_args[LoraConfig-config_kwargs7-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpt9odnm0x/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpt9odnm0x/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate_pos_args[LoraConfig-config_kwargs8-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpdogclf3u/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpdogclf3u/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate_pos_args[VBLoRAConfig-config_kwargs14-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._dynamo.exc.Unsupported: Data-dependent branching
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate_half_prec[LoraConfig-config_kwargs7-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpey5wj39y/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpey5wj39y/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_generate_half_prec[LoraConfig-config_kwargs8-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpl6j7lpr_/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpl6j7lpr_/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_disable_adapter[AdaLoraConfig-config_kwargs0-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpsfxc4jg3/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpsfxc4jg3/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_disable_adapter[LoraConfig-config_kwargs7-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpwes5jor3/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpwes5jor3/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_disable_adapter[LoraConfig-config_kwargs8-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmpqbmim7eh/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmpqbmim7eh/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
FAILED tests/test_decoder_models.py::TestDecoderModels::test_disable_adapter[VBLoRAConfig-config_kwargs14-hf-internal-testing/tiny-random-Gemma3ForCausalLM] - torch._inductor.exc.InductorError: CalledProcessError: Command '['/usr/bin/gcc', '/tmp/tmp0xr7hvk5/main.c', '-O3', '-shared', '-fPIC', '-o', '/tmp/tmp0xr7hvk5/cuda_utils.cpython-310-x86_64-linux-gnu.so', '-lcuda', '-L/usr/local/lib/python3.10/dist-packages/triton/backends/nvidia/lib', '-L/usr/lib/x86_64-li...
================================================================================================================ 16 failed, 842 passed, 296 skipped, 3480 deselected, 374 warnings in 533.36s (0:08:53) ================================================================================================================
```

> 
> > I had forced the dtype of x and delta_weight to become torch.float32 (and that was still throwing the same error on our end). I am not sure if that affected your tests in anyway.
> 
> I removed the coercion to float32 and then I get the `"addmm_sparse_cuda" not implemented for 'BFloat16'`. However, conversion fixes it for me.

We have identified the issue on our end for this addmm_sparse_cuda error. Specifically, we were setting bf16 training arguments flag inside SFTTrainer/Transformers trainer and that was resulting in that error for bfloat16. Removing this flag resulted in autocast_adapter_dtype working as expected and all errors disappeared. 

We have now made this new forward pass as our main SHiRA implementation. Thanks for the suggestion. One thing we wanted to point out for the new implementation: We ran some simple profiling experiments for LoRA, new SHiRA implementation, and old SHiRA implementation and here are the results (OPT-125m model and r = 16, A100 GPU, same setup for torch and transformers as above):

```
# LoRA: 2,359,296 params
5857MB GPU memory as per NVIDIA-SMI, 'train_runtime': 12.1698, 'train_samples_per_second': 52.589, 'train_steps_per_second': 1.643.

# SHiRA new forward: 2,359,296 parameters
5679MB GPU memory as per NVIDIA-SMI, 'train_runtime': 24.4731, 'train_samples_per_second': 26.151, 'train_steps_per_second': 0.817.

# SHiRA old forward: 2,359,296 parameters
4679MB GPU memory as per NVIDIA-SMI, 'train_runtime': 10.8753, 'train_samples_per_second': 58.849, 'train_steps_per_second': 1.839.
```

It appears that our original implementation was the fastest and was also the most memory-efficient. This is likely because we were avoiding a secondary branch for the adapters that would reduce the activation memory and also the new operations on the adapter branch would not be needed. Also, the linear with sparse_coo_tensor seems quite slow compared to our original implementation that avoided this sparse linear operation. 

Any ideas on how to improve upon this? I understand that the new implementation makes it more generic in case someone is subclassing the nn.Linear layers in their base model. However, any training speed improvement ideas would be appreciated. We are also open to keeping both forward implementations using some kind of a config parameter, if that is acceptable. Please let us know. Thanks.

BenjaminBossan commented:
Thanks for the updates.

> It appears that our original implementation was the fastest and was also the most memory-efficient.

Thanks for testing, that was definitely a good idea. I did some of my own testing using a bigger model, `meta-llama/Llama-3.2-3B`, using the [PEFT method comparison suite](https://github.com/huggingface/peft/tree/main/method_comparison/MetaMathQA). I trained for 500 steps and got these results:

| Metric           | Old      | New     |
|------------------|---------------|---------------|
| cuda_memory_max  | 19046334464   | 19761463296   |
| valid accuracy   | 0.42          | 0.42          |
| train loss       | 0.752         | 0.752         |
| train time (s)   | 29.9          | 87.1          |
| eval time (s)    | 10.0          | 14.3          |
| tokens / sec     | 6951.8        | 2387.2        |

So I can confirm that the new forward method is quite a bit slower and requires a little bit more memory. This was surprising to me at first, so I dug a bit deeper.

> This is likely because we were avoiding a secondary branch for the adapters that would reduce the activation memory and also the new operations on the adapter branch would not be needed.

I'm not quite sure if that's what you mean, but what I found is that `F.linear(x, delta_weight)` is the culprit, as there is apparently no highly efficient implementation for sparse COO matrices and since the allocated activations can actually be quite large. I tried a couple of things (e.g. with and without compile) but ultimately, nothing helped.

Thus I think going back to the original version is the right choice. It has a other few disadvantages, but I think we can live with that given how much faster it is. Still could you please add the following sanity check to `shira.Linear.__init__`:

```python
        ShiraLayer.__init__(self, base_layer, **kwargs)
        self.fan_in_fan_out = fan_in_fan_out
        if self.base_layer is not self.get_base_layer():
            raise ValueError("SHiRA does not support nested base layers")
```

This should help mitigate a potential source of errors. Maybe some tests need adjusting after this, not sure.

> However, for this latest peft-gpu setup, these errors are not specific to SHiRA. VeRA and LoRA tests also fail for the Gemma model with similar errors

Could you please merge with, or rebase on, the latest PEFT `main`? This should hopefully resolve the Gemma errors. There are some merge conflicts but those should be easy to fix.

kkb-code commented:
> I'm not quite sure if that's what you mean, but what I found is that F.linear(x, delta_weight) is the culprit, as there is apparently no highly efficient implementation for sparse COO matrices and since the allocated activations can actually be quite large. I tried a couple of things (e.g. with and without compile) but ultimately, nothing helped.
>
> Thus I think going back to the original version is the right choice. It has a other few disadvantages, but I think we can live with that given how much faster it is. Still could you please add the following sanity check to `shira.Linear.__init__`:
> ```
>        ShiraLayer.__init__(self, base_layer, **kwargs)
>        self.fan_in_fan_out = fan_in_fan_out
>        if self.base_layer is not self.get_base_layer():
>            raise ValueError("SHiRA does not support nested base layers")
> ```

Thanks for confirming and for running new tests. Yes, this is exactly what I meant. We have now reverted back to the original forward implementation and have added the sanity check you suggested for nested base layers.

> Could you please merge with, or rebase on, the latest PEFT main? This should hopefully resolve the Gemma errors. There are some merge conflicts but those should be easy to fix.

We have merged PEFT main into the SHiRA pull request. All the merge conflicts were resolved. However, we are still seeing the Gemma errors (same 3 tests fail as before for SHiRA, VeRA, LoRA, and even for the newly added C3A adapter) with the latest transformers setup. We think this is not SHiRA-specific, so please let us know if any action is needed from our end to resolve this issue.

Finally, we have added the benchmark configs for SHiRA under `method_comparison/MetaMathQA/experiments/shira/llama-3.2-3B-default/`. Please let us know if any changes are needed here. 

Thanks a lot. Please let us know if our pull request is ready to merge.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_2584). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BenjaminBossan reviewed (changes requested):
> Thanks for confirming and for running new tests. Yes, this is exactly what I meant. We have now reverted back to the original forward implementation and have added the sanity check you suggested for nested base layers.

Thanks for that.

> However, we are still seeing the Gemma errors (same 3 tests fail as before for SHiRA, VeRA, LoRA, and even for the newly added C3A adapter)

I checked again. What appears to be happening is that these errors do indeed occur, but only on GPU. On CPU, these tests pass. The CI uses only CPU, so I think for now we can ignore this. I opened an issue to look at this later.

> Finally, we have added the benchmark configs for SHiRA under `method_comparison/MetaMathQA/experiments/shira/llama-3.2-3B-default/`. Please let us know if any changes are needed here.

Thanks, I ran the experiment locally. Generally, results seem to be close to vanilla LoRA with rank 32:

| Metric                     | LoRA           | SHiRA          |
|----------------------------|----------------|----------------|
| cuda_memory_reserved_avg   | 11868689976    | 12262929386    |
| cuda_memory_max            | 22273851392    | 21705523200    |
| cuda_memory_reserved_99th  | 17710763212    | 17666618163    |
| file_size (bytes)          | 36715216       | 110115520      |
| num_trainable_params       | 9175040        | 9175040        |
| train loss                 | 0.607          | 0.579          |
| test accuracy              | 0.482          | 0.519          |

(I omitted the training time, as the two were tested on different machines. Final results will follow once merged)

Do these results look correct and reasonable to you? Most surprising is perhaps the file size difference, I wonder where that comes from, after a quick check, it's not the dtypes (all float32 or int32).

> Thanks a lot. Please let us know if our pull request is ready to merge.

Apart from my small comments, could you please always run `make style` before committing to satisfy the linter?

kkb-code commented on method_comparison/MetaMathQA/experiments/shira/llama-3.2-3B-default/adapter_config.json:
Thanks. This is done. I have renamed the folder to `llama-3.2-3B-lr_0.0003-random_seed_42` because I was also fixing the random seed to 42 (just to make sure the results are reproducible).

kkb-code commented:
> Thanks, I ran the experiment locally. Generally, results seem to be close to vanilla LoRA with rank 32:
> 
> Metric	LoRA	SHiRA
> cuda_memory_reserved_avg	11868689976	12262929386
> cuda_memory_max	22273851392	21705523200
> cuda_memory_reserved_99th	17710763212	17666618163
> file_size (bytes)	36715216	110115520
> num_trainable_params	9175040	9175040
> train loss	0.607	0.579
> test accuracy	0.482	0.519
> (I omitted the training time, as the two were tested on different machines. Final results will follow once merged)
> 
> Do these results look correct and reasonable to you? Most surprising is perhaps the file size difference, I wonder where that comes from, after a quick check, it's not the dtypes (all float32 or int32).

These results look correct and are completely reasonable. The file size difference is because our SHiRA checkpoint saves both the SHiRA weights as well as the indices used during the training run. When loading the checkpoint, the indices are also loaded along with the adapter weights. All these necessary features are implemented in the `save_and_load.py` file. Saving and loading the indices makes sure that users are loading the exact same mask that they trained with. Of note, we are saving the indices in int32 (that was the minimum int dtype we could use for sparse_coo_tensor) instead of int64 to reduce the adapter checkpoint size.

> Apart from my small comments, could you please always run `make style` before committing to satisfy the linter?

Sorry, I had missed the `doc-builder` part of `make style` earlier. Should be fixed now.

kkb-code commented:
> This comment is still open. I think it would be really great to support the other mask types too. It would also be fine to add them in a subsequent PR.

About the other mask_types, currently, we have not planned for this. Perhaps we can consider this in a subsequent PR. The current PR is more than sufficient for users to adopt SHiRA, particularly with our custom_mask_function support. Hence, I think we can proceed with merging this PR (if everything looks acceptable). Thanks so much for the guidance throughout this process. Please let us know if you need anything further.

kkb-code commented on src/peft/tuners/shira/model.py:
```diff
@@ -0,0 +1,331 @@
+        >>> from peft import PeftModel
+
+        >>> base_model = AutoModelForCausalLM.from_pretrained("tiiuae/falcon-40b")
+        >>> peft_model_id = "smangrul/falcon-40B-int4-peft-lora-sfttrainer-sample"
```
Fixed.

kkb-code commented on src/peft/utils/save_and_load.py:
```diff
@@ -152,6 +152,14 @@ def renamed_dora_weights(k):
+        for name, module in model.named_modules():
+            if hasattr(module, "shira_indices"):
+                for k, v in module.shira_indices.items():
+                    to_return[f"{name}.shira_indices.{k}"] = v
```
> I'm trying to understand this part. IIUC, it is about bringing the SHiRA indices into the `state_dict`. They are not part yet because they're just in a normal dict. Could we not turn that dict into a `nn.ParameterDict` and thus the indices would be included automatically?

We had originally tried to make the shira_indices nn.ParameterDict. The problem was that despite setting the correct dtype (int) when we create those ParameterDicts, autocast_adapter_dtype turns them into a float and then sparse_coo_tensor fails (because the indices must be integers). This is why we opted to make those indices normal dicts and then save them separately. 

> Also, for testing, I removed this code completely and for me, all SHiRA tests still passed (`pytest tests/ -k shira`). So either this code in unnecessary, which I don't think is the case, or the test coverage is insufficient. I'm just guessing here, but is it possible that in the tests, the indices are restored correctly even without this code because of the random seed? Anyway, I think we need to investigate this.

We have now added a new test which would fail if you remove the SHiRA indices saving part from `save_and_load.py`. This saving and loading part is especially critical when we are using custom masks. We have improved the code further so that the indices only get loaded from the checkpoint when we are using `from_pretrained` method. Also, for the cases where no custom_mask is used (added another test for this), we saw that torch.Generator seems to have its own random_seed even when we are not fixing it. This might be why the tests were passing previously. In any case, now the indices are instantiated only from the checkpoint (earlier, they were getting initialized by a random mask and then overwritten by the checkpoint later). So this should fix things.

kkb-code commented on tests/test_shira.py:
```diff
@@ -0,0 +1,175 @@
+        assert torch.all(output_first - output_first_loaded == 0)
+        assert torch.all(output_second - output_second_loaded == 0)
+
+        assert torch.all(shira_assign_val1_f - peft_model.base_model.model.lin1.shira_weight["first"] == 0)
```
Done.

kkb-code commented on tests/test_shira.py:
```diff
@@ -0,0 +1,175 @@
+        peft_model.set_adapter("second")
+        output_second_loaded = peft_model(input)
+
+        assert torch.all(output_first - output_first_loaded == 0)
```
Done.

kkb-code commented on tests/test_custom_models.py:
```diff
@@ -1492,6 +1526,12 @@ def test_forward_float16_no_autocast(self, test_name, model_id, config_cls, conf
+        if config_cls in [
+            ShiraConfig,
+        ]:
+            pytest.skip(
```
Done.

kkb-code commented on tests/test_custom_models.py:
```diff
@@ -1533,6 +1573,12 @@ def test_forward_bfloat16_no_autocast(self, test_name, model_id, config_cls, con
+        if config_cls in [
+            ShiraConfig,
+        ]:
+            pytest.skip(
```
Done.

kkb-code commented on src/peft/tuners/shira/layer.py:
```diff
@@ -0,0 +1,213 @@
+        self.r = {}
+        self.scaling = {}
+        self.shira_weight = nn.ParameterDict({})
+        self.shira_indices = {}
```
Done.

kkb-code commented:
> I did another review round and mostly, everything looks good but I did stumble across some smaller issues. It's not much, but let's deal with those and then we should be ready to merge the PR.

Thanks for pointing these out. We have added 4 new commits that address all of these issues. Thanks again, please let us know if you need anything else.

BenjaminBossan commented on tests/test_shira.py:
```diff
@@ -0,0 +1,284 @@
+
+
+def custom_random_mask_function_with_custom_kwargs(custom_arg_1, custom_arg_2):
+    print(
```
Remove

BenjaminBossan commented on tests/test_shira.py:
```diff
@@ -0,0 +1,284 @@
+from peft import PeftModel, ShiraConfig, get_peft_model
+
+
+def custom_random_mask_function_with_custom_kwargs(custom_arg_1, custom_arg_2):
```
I don't quite get the point of having two args here, would one not suffice?

BenjaminBossan commented on tests/test_shira.py:
```diff
@@ -0,0 +1,284 @@
+        # so it will lead to different random sparse masks between saving and loading.
+        # our goal is to make sure that loaded indices are exactly the same as the saved indices regardless of what initial random mask gets generated.
+        # we will also make sure that parameters are saved and loaded correctly, and the output remains the same.
+        # torch.manual_seed(1)
```
Remove?

kkb-code replied:
Done.

BenjaminBossan reviewed (changes requested):
Thanks for the latest updates. I just have a few small comments, otherwise we're good to go.

> We had originally tried to make the shira_indices nn.ParameterDict. The problem was that despite setting the correct dtype (int) when we create those ParameterDicts, autocast_adapter_dtype turns them into a float and then sparse_coo_tensor fails (because the indices must be integers). This is why we opted to make those indices normal dicts and then save them separately.

It would be strange if `autocast_adapter_dtype` had that effect, as we explicitly check the dtype. I did some tests and the issue that I came across was instead that there is an error because int dtype cannot have gradients and it is automatically assumed that all `nn.Parameter` can have gradients. I tried to disable it but it's not straightforward (or impossible?) so it's not worth the effort. We could use [`BufferDict`](https://github.com/huggingface/peft/blob/45996a1d6e00b0951424a210d24aa4058c7b0dcf/src/peft/tuners/_buffer_dict.py#L19) with `persistent=True`. This could simplify the code in `save_and_load.py` since the buffers should automatically be included.

As always, once you finished your changes, please call `make style`.

kkb-code commented:
> Thanks for the latest updates. I just have a few small comments, otherwise we're good to go.
> 
> > We had originally tried to make the shira_indices nn.ParameterDict. The problem was that despite setting the correct dtype (int) when we create those ParameterDicts, autocast_adapter_dtype turns them into a float and then sparse_coo_tensor fails (because the indices must be integers). This is why we opted to make those indices normal dicts and then save them separately.
> 
> It would be strange if `autocast_adapter_dtype` had that effect, as we explicitly check the dtype. I did some tests and the issue that I came across was instead that there is an error because int dtype cannot have gradients and it is automatically assumed that all `nn.Parameter` can have gradients. I tried to disable it but it's not straightforward (or impossible?) so it's not worth the effort. We could use [`BufferDict`](https://github.com/huggingface/peft/blob/45996a1d6e00b0951424a210d24aa4058c7b0dcf/src/peft/tuners/_buffer_dict.py#L19) with `persistent=True`. This could simplify the code in `save_and_load.py` since the buffers should automatically be included.
> 
> As always, once you finished your changes, please call `make style`.

We attempted the `BufferDict` method for `shira_indices` as documented in [this commit](https://github.com/huggingface/peft/pull/2584/commits/8c68a5a5ddf8039aa3340652db8c226a11c086d5). We reverted back to our original method since the `BufferDict` method did not work. 

The problems are as follows: 

1. If we use the `BufferDict` method to declare the empty shira_indices dictionary, the adapter loading fails in the case of custom mask functions because when we call the `from_pretrained` method, the specific keys where we want to load the `shira_indices` are absent. This throws a `KeyError`.
2. We rectified the above issue by creating the keys explicitly in the case of custom mask functions but we set them to `None` [here](https://github.com/huggingface/peft/pull/2584/commits/8c68a5a5ddf8039aa3340652db8c226a11c086d5#diff-f8f796a371f99323586a2eb8124cb741f7eb57ff6df6864b4d5aa54935a8c1f2R102) (because when we use `from_pretrained` to load the adapter checkpoints, `config.mask_fn = None`, so we expect the correct indices to be loaded directly from the checkpoint). Please see point 3 below for why we want these indices to be None in case of a custom_mask_function (if no `config.mask_fn` is specified). Anyways, in this case, torch starts identifying the `shira_indices` inside the pretrained checkpoint as `unexpected_keys` and does not load the indices on top of their initial `None` values.  Hence, the inference fails in the forward pass.
3. The only way where the shira_indices loading worked correctly was to declare an empty tensor using `torch.empty` that gets the correct `dtype` and `shape` (this line is [here](https://github.com/huggingface/peft/pull/2584/commits/8c68a5a5ddf8039aa3340652db8c226a11c086d5#diff-f8f796a371f99323586a2eb8124cb741f7eb57ff6df6864b4d5aa54935a8c1f2R101)) However, this introduces a serious bug during the training time in case someone forgets to provide `config.mask_fn` for a custom mask function case. The issue here is that normally, if you do not provide a `mask_fn` for the custom case, SHiRA training would just throw an error since `shira_indices = {}`. But if we declare them to be an empty tensor, torch populates them with random integer values and `torch.sparse_coo_tensor` does _not_ throw any errors (we verified this). This means that the model would appear to train, but the indices would be incorrect or random. 

Obviously, we need to avoid all these issues, so we reverted back to the original stable implementation.

kkb-code commented on src/peft/tuners/shira/config.py:
```diff
@@ -137,6 +123,6 @@ def __post_init__(self):
         else:
             warnings.warn(
-                f"Argument {self.mask_type=} is not recognized, please supply your own masking function by calling `config.mask_fn = my_mask_fn`."
+                f"Argument {self.mask_type=} is not recognized, please supply your own masking function by calling `config.mask_fn = my_mask_fn`. Please ignore this warning message if you are loading from a pretrained checkpoint"
```
I cannot think of anything that would fix this minor issue. We need that warning to make sure users understand that during training, they must set `config.mask_fn` in a meaningful way (for the custom case), otherwise the training will throw an error. However, when the custom mask function trained adapter is loaded using `from_pretrained`, `config.mask_fn` will be set to `None` (from the checkpoint) and, therefore, this warning will be displayed. So, I don't really see a way around it.

kkb-code commented on examples/shira_finetuning/shira_sft_finetuning.py:
Deleted.

BenjaminBossan commented on src/peft/tuners/shira/config.py:
```diff
@@ -0,0 +1,128 @@
+        if self.mask_type == "random":
+            self.mask_fn = random_mask
+        else:
+            warnings.warn(
+                f"Argument {self.mask_type=} is not recognized, please supply your own masking function by calling `config.mask_fn = my_mask_fn`. Please ignore this warning message if you are loading from a pretrained checkpoint."
+            )
```
I think this should be a good way to avoid showing the warning if not needed:

```suggestion
            if not self.inference_mode:
                warnings.warn(
                    f"Argument {self.mask_type=} is not recognized, please supply your own masking function by calling "
                    "`config.mask_fn = my_mask_fn`. Please ignore this warning message if you are loading from a "
                    "pretrained checkpoint."
                )
```

PEFT automatically sets `inference_mode` on the config when loading a PEFT model:

https://github.com/huggingface/peft/blob/35000fda88fdf29db785783f8a40431cee2cab10/src/peft/peft_model.py#L450

kkb-code replied:
Very good suggestion. We have added this now and it works. Thanks.

BenjaminBossan reviewed (changes requested):
Thanks for your amendments and trying out if `BufferDict` would help with the persistence code. Too bad this causes new issues, I agree that sticking with the first approach is the best way forward.

I investigated a bit more how we can avoid the spurious warning about `mask_type` and I believe my solution should help with that. Please check my suggestion.

Please also merge with/rebase on the latest main, as there was an issue with the latest transformers release. Then, CI should pass and we should be good to merge.

kkb-code commented:
> I investigated a bit more how we can avoid the spurious warning about `mask_type` and I believe my solution should help with that. Please check my suggestion.

Added.

> Please also merge with/rebase on the latest main, as there was an issue with the latest transformers release. Then, CI should pass and we should be good to merge.

We have now merged with the latest main. We also verified `make_style` and everything passed. This should be good to go from our side. Thanks a lot.

BenjaminBossan commented:
This is indeed a very strange situation and not something I've come across yet. Today, I re-ran the CI and still the same errors occur. Too bad that we don't have Windows machines to test on, so we need to resort to the GH CI instead.

Perhaps we can try to eliminate potential sources of error one by one. The first thing I'd test is to comment out all the additions to `save_and_load.py`. Of course, this will lead some tests to fail, but I'm curious if that's enough to fix the Windows permission error. Could you please try that? In addition, to avoid overexerting the CI, let's only test on a single Python version:

https://github.com/huggingface/peft/blob/35000fda88fdf29db785783f8a40431cee2cab10/.github/workflows/tests.yml#L45

```diff
-       python-version: ["3.9", "3.10", "3.11", "3.12"]
+       python-version: ["3.12"]
```

Of course, we will revert these changes before merging.

kkb-code commented:
Interestingly, when I remove all SHiRA related lines from `save_and_load.py`, windows does not throw the `PermissionError`. Not sure what is going on...but I added another commit reverting back the `save_and_load.py` to see if it still fails. The windows and ubuntu tests on commit [34af1fd](https://github.com/huggingface/peft/pull/2584/commits/34af1fdc572d333526fc25a8cca5b193f35e3654) failed only 1 test which is expected (the custom mask save and load test would fail if we do not save and load shira-indices for the custom mask).

BenjaminBossan commented:
> Any help would be appreciated. Thanks.

I'm as stumped as you are, this really doesn't make a lot of sense. I wondered if you could test converting the ints to floats before saving and then back to ints when loading. This is not a final solution, just for testing if it could somehow be related to the dtype.

kkb-code commented:
Just pushed a commit for SHiRA indices dtype changed to torch.float32 while saving and then torch.int while loading. Let us see if this helps. 

Can we think of any other solutions that might help us merge this? For example, can we skip certain tests on windows OS only? This clearly seems like a Permission issue that happens only on the windows. If there was anything wrong with our code, we would see these issues happening on other OS too?

BenjaminBossan commented:
> Just pushed a commit for SHiRA indices dtype changed to torch.float32 while saving and then torch.int while loading. Let us see if this helps.

This seems to help indeed, there is only one [Windows error](https://github.com/huggingface/peft/actions/runs/16221180966/job/45802927087?pr=2584) related to 429 from the Hub, so it's good.

> can we skip certain tests on windows OS only? This clearly seems like a Permission issue that happens only on the windows. If there was anything wrong with our code, we would see these issues happening on other OS too?

While generally I agree, I think there is a possibility that something more fundamental is broken here, so before we don't fully understand what is going on, merging is risky. I can't imagine that saving integers in the `state_dict` on Windows is generally broken, but maybe something more subtle is going on. Unfortunately, I currently don't have time to investigate and don't have a Windows machine, but perhaps you have an idea based on the latest finding.

kkb-code commented:
> This seems to help indeed, there is only one [Windows error](https://github.com/huggingface/peft/actions/runs/16221180966/job/45802927087?pr=2584) related to 429 from the Hub, so it's good.
> While generally I agree, I think there is a possibility that something more fundamental is broken here, so before we don't fully understand what is going on, merging is risky. I can't imagine that saving integers in the `state_dict` on Windows is generally broken, but maybe something more subtle is going on. Unfortunately, I currently don't have time to investigate and don't have a Windows machine, but perhaps you have an idea based on the latest finding.

Okay, this is good, at least. Actually, it seems that there is something going on with Integers and Safetensors that is specific only to the Windows OS. Someone reported another issue on Safetensors github: [please see here](https://github.com/huggingface/safetensors/issues/391#issue-2002307008). Potentially looks like some kind of an overflow problem in Windows. 

> The reported array length is exactly 2 * (test_tensor.numel() - 2**32), leading me to suspect this is an integer overflow issue. The issue only seems to manifest on Windows, with the code above performing as expected on Linux.

Seems like the above issue was closed as it was not planned, so I don't think any fix was provided.

I am okay with casting indices to float32 while saving and then casting them back to int when loading them as a solution to this. Seems like it does not affect any Linux tests.

BenjaminBossan commented:
> I am okay with casting indices to float32 while saving and then casting them back to int when loading them as a solution to this. Seems like it does not affect any Linux tests.

Okay, so how about only doing that when the platform is Windows? Then we can be sure that Linux and MacOS are unaffected. For reasonably small values, this should work fine, but at a certain size, the conversion should fail. I wonder if this point is known in advance.

kkb-code commented:
I agree, precision issues can occur for very large integer values. So let me just limit this to windows systems. I will use something like `platform.system() == 'Windows'`. Similar `platform.system()` statements have been used in other places inside peft so this should work. Will push shortly.

kkb-code commented:
I just pushed [cc47b6a](https://github.com/huggingface/peft/pull/2584/commits/cc47b6a54e0378628ef43de100fc63d59993383c). It brings back all tests (all python versions) and casts the shira_indices to float32 before saving only on windows. Hence, Mac-OS and Linux must be completely unaffected now. Thanks. Let us see if this works. 

BTW, I am not sure what to do about Hub errors. I thought those were not happening before.

kkb-code commented:
Yep, thanks. Just pushed another commit. This should take care of the 120 char limit.

BenjaminBossan reviewed (approved):
Thanks for taking care of the Windows issue and this contribution in general. The failing MacOS test is just flakiness and can be ignored. The PR is good to be merged.

kkb-code commented:
Thank you so much @BenjaminBossan for the guidance throughout this process. We are happy to see that our SHiRA adapter was merged into PEFT. Please let us know if we should close Issue #2585 or if you would close it at a later time. 

Also, I am wondering if there is any ETA on the next PEFT version release that would include the SHiRA update (e.g., v0.16.1)? Can you please let us know? That would allow users to directly use `pip install ...` and be able to use SHiRA. 

Thanks again for the help. It was great working with you on this.

BenjaminBossan commented:
> Also, I am wondering if there is any ETA on the next PEFT version release that would include the SHiRA update (e.g., v0.16.1)? Can you please let us know? That would allow users to directly use `pip install ...` and be able to use SHiRA.

Even though we had a release just recently, we might have another one soon, but we have external dependencies for that, so I can't give a date. FYI, for new features such as SHiRA, we always do minor releases (i.e. it would be 0.17.0), patch releases are for hot fixes only. In the meantime, users can install directly from GH, e.g. `pip install git+https://github.com/huggingface/peft` (+ optional hash to fix the commit).

> Thanks again for the help. It was great working with you on this.

Thanks, same from my side.
